src/components/timeline/Row.svelte

```diff
@@ -362,72 +395,114 @@
-      // if more than one layer matches a type
-      let seenDirectiveIds: Record<number, boolean> = {};
-      let seenSpanIds: Record<number, boolean> = {};
-      activityLayers.forEach(layer => {
-        if (layer.filter && layer.filter.activity !== undefined) {
-          const types = layer.filter.activity.types || [];
-          types.forEach(type => {
-            const matchingDirectives = directivesByType[type];
-            if (matchingDirectives) {
-              const uniqueDirectives: ActivityDirective[] = [];
-              matchingDirectives.forEach(directive => {
-                if (!seenDirectiveIds[directive.id]) {
-                  idToColorMaps.directives[directive.id] = layer.activityColor;
-                  seenDirectiveIds[directive.id] = true;
-                  uniqueDirectives.push(directive);
-                }
-              });
-              directives = directives.concat(uniqueDirectives);
-            }
-            const matchingSpans = spansByType[type];
-            if (matchingSpans) {
-              const uniqueSpans: Span[] = [];
-              matchingSpans.forEach(span => {
-                if (!seenSpanIds[span.span_id]) {
-                  idToColorMaps.spans[span.span_id] = layer.activityColor;
-                  seenSpanIds[span.span_id] = true;
-                  uniqueSpans.push(span);
-                }
-              });
-              spans = spans.concat(uniqueSpans);
+  $: if (
+    (activityLayers && spansMap && activityDirectives && typeof filterItemsByTime === 'boolean') ||
+    hasExternalEventsLayer
+  ) {
+    discreteTree = [];
+    if (activityLayers && spansMap && activityDirectives && typeof filterItemsByTime === 'boolean') {
+      idToColorMaps.directives = {};
+      idToColorMaps.spans = {};
+
+      let spansList = Object.values(spansMap);
+      const directivesByType = groupBy(activityDirectives, 'type');
+      const spansByType = groupBy(spansList, 'type');
+      if (activityLayers.length) {
+        let directives: ActivityDirective[] = [];
+        let spans: Span[] = [];
+
+        // track directives and spans that have been seen to avoid double counting
+        // if more than one layer matches a type
+        let seenDirectiveIds: Record<number, boolean> = {};
+        let seenSpanIds: Record<number, boolean> = {};
+        activityLayers.forEach(layer => {
+          if (layer.filter && layer.filter.activity !== undefined) {
+            const types = layer.filter.activity.types || [];
+            types.forEach(type => {
+              const matchingDirectives = directivesByType[type];
+              if (matchingDirectives) {
+                const uniqueDirectives: ActivityDirective[] = [];
+                matchingDirectives.forEach(directive => {
+                  if (!seenDirectiveIds[directive.id]) {
+                    idToColorMaps.directives[directive.id] = layer.activityColor;
```

Comment:
Would it be possible to not mutate anything variables that are props for components? Mutating causes svelte to not always know when to rerender. I suggest completely rebuilding `idToColorMaps` rather than setting member variables to ensure all the components depending on it update correctly.

Reply:
Addressed [here](https://github.com/NASA-AMMOS/aerie-ui/pull/1396/commits/1c467d0f10c8357359300e093895fd58f137796a)!